src/WebSocketStream.ts

```diff
@@ -1,5 +1,523 @@
+    // Force close any open streams
+    await this.cancel(new errors.ErrorWebSocketStreamClose());
+    // Removing stream from the connection's stream map
+    // TODO: the other side currently will send back an ERROR/CLOSE frame from us sending an ERROR/CLOSE frame from this.close().
+    // However, out stream gets deleted before we receive that message on the connection.
+    // So the connection will infinitely create streams with the same streamId when it receives the ERROR/CLOSE frame.
+    // I'm dealing with this by just filtering out ERROR/CLOSE frames in the connection's onMessage handler, but there might be a better way to do this.
+    this.connection.streamMap.delete(this.streamId);
+    this.logger.info(`Destroyed ${this.constructor.name}`);
+  }
+
+  protected async streamSend(message: StreamMessage): Promise<void> {
+    const array: Array<Uint8Array> = [
+      this.encodedStreamId,
+      ...generateStreamMessage(message, false),
+    ];
+    await this.connection.send(array);
+  }
+
+  /**
+   * Put a message frame into a stream.
+   * This will not will not error out, but will rather close the ReadableStream assuming any further reads are expected to fail.
+   * @param message - The message to put into the stream.
+   * @internal
+   */
+  public async streamRecv(message: Uint8Array) {
+    if (message.length === 0) {
+      this.logger.debug(`received empty message, closing stream`);
+      await this.signalReadableEnd(
+        true,
+        new errors.ErrorWebSocketStreamReadableParse('empty message', {
+          cause: new RangeError(),
+        }),
+      );
+      return;
+    }
+
+    let parsedMessage: StreamMessage;
+    try {
+      parsedMessage = parseStreamMessage(message);
+    } catch (err) {
+      await this.signalReadableEnd(
+        true,
+        new errors.ErrorWebSocketStreamReadableParse(err.message, {
+          cause: err,
+        }),
+      );
+      return;
+    }
+
+    if (parsedMessage.type === StreamMessageType.Ack) {
+      this.writableDesiredSize += parsedMessage.payload;
+      this.writableDesiredSizeProm.resolveP();
+      this.logger.debug(
+        `writableDesiredSize is now ${this.writableDesiredSize} due to ACK`,
+      );
+    } else if (parsedMessage.type === StreamMessageType.Data) {
+      if (this._readableEnded) {
+        return;
+      }
+      if (
+        parsedMessage.payload.length >
+        this.readableQueueBufferSize - this.readableQueue.length
+      ) {
+        await this.signalReadableEnd(
+          true,
+          new errors.ErrorWebSocketStreamReadableBufferOverload(),
+        );
+        return;
+      }
+      this.readableQueue.queue(parsedMessage.payload);
+      this.readableBufferReady.resolveP();
+    } else if (parsedMessage.type === StreamMessageType.Error) {
+      const { shutdown, code } = parsedMessage.payload;
+      let reason: any;
+      switch (code) {
+        case BigInt(StreamErrorCode.Unknown):
```

Comment:
You can use regular constants like `StreamErrorCode = { Unknown: 123n, ... }`.

That avoids having to keep converting them to `BigInt` all the time.

Also switch case matches is always `===`.

Reply:
```
> switch(o) {
... case o:
...   console.log('yes');
... }
yes
```

Reply:
Changed to:
```ts
const StreamErrorCode = {
  Unknown: 0n as VarInt,
  ErrorReadableStreamParse: 1n as VarInt,
  ErrorReadableStreamBufferOverflow: 2n as VarInt,
} as const;
```